drivers/mspi/mspi_dw.c

```diff
@@ -0,0 +1,1154 @@
+					  XIP_WRITE_CTRL_FRF_QUAD);
+		break;
+	case MSPI_IO_MODE_OCTAL:
+	case MSPI_IO_MODE_OCTAL_1_1_8:
+	case MSPI_IO_MODE_OCTAL_1_8_8:
+		ctrl->read |= FIELD_PREP(XIP_CTRL_FRF_MASK,
+					 XIP_CTRL_FRF_OCTAL);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_FRF_MASK,
+					  XIP_WRITE_CTRL_FRF_OCTAL);
+		break;
+	default:
+		LOG_ERR("IO mode %d not supported", io_mode);
+		return false;
+	}
+
+	/* Transfer format used for Address and Instruction: */
+
+	switch (io_mode) {
+	case MSPI_IO_MODE_DUAL_1_1_2:
+	case MSPI_IO_MODE_QUAD_1_1_4:
+	case MSPI_IO_MODE_OCTAL_1_1_8:
+		/* - both sent in Standard SPI mode */
+		ctrl->read |= FIELD_PREP(XIP_CTRL_TRANS_TYPE_MASK,
+					 XIP_CTRL_TRANS_TYPE_TT0);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_TRANS_TYPE_MASK,
+					  XIP_WRITE_CTRL_TRANS_TYPE_TT0);
+		break;
+	case MSPI_IO_MODE_DUAL_1_2_2:
+	case MSPI_IO_MODE_QUAD_1_4_4:
+	case MSPI_IO_MODE_OCTAL_1_8_8:
+		/* - Instruction sent in Standard SPI mode,
+		 *   Address sent the same way as data
+		 */
+		ctrl->read |= FIELD_PREP(XIP_CTRL_TRANS_TYPE_MASK,
+					 XIP_CTRL_TRANS_TYPE_TT1);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_TRANS_TYPE_MASK,
+					  XIP_WRITE_CTRL_TRANS_TYPE_TT1);
+		break;
+	default:
+		/* - both sent the same way as data. */
+		ctrl->read |= FIELD_PREP(XIP_CTRL_TRANS_TYPE_MASK,
+					 XIP_CTRL_TRANS_TYPE_TT2);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_TRANS_TYPE_MASK,
+					  XIP_WRITE_CTRL_TRANS_TYPE_TT2);
+		break;
+	}
+
+	return true;
+}
+
+static bool apply_cmd_length(struct mspi_dw_data *dev_data, uint32_t cmd_length)
+{
+	switch (cmd_length) {
+	case 0:
+		dev_data->spi_ctrlr0 |= FIELD_PREP(SPI_CTRLR0_INST_L_MASK,
+						   SPI_CTRLR0_INST_L0);
+		break;
+	case 1:
+		dev_data->spi_ctrlr0 |= FIELD_PREP(SPI_CTRLR0_INST_L_MASK,
+						   SPI_CTRLR0_INST_L8);
+		break;
+	case 2:
+		dev_data->spi_ctrlr0 |= FIELD_PREP(SPI_CTRLR0_INST_L_MASK,
+						   SPI_CTRLR0_INST_L16);
+		break;
+	default:
+		LOG_ERR("Command length %d not supported", cmd_length);
+		return false;
+	}
+
+	return true;
+}
+
+static bool apply_xip_cmd_length(const struct mspi_dw_data *dev_data,
+				 struct xip_ctrl *ctrl)
+{
+	uint8_t cmd_length = dev_data->xip_params_active.cmd_length;
+
+	switch (cmd_length) {
+	case 0:
+		ctrl->read |= FIELD_PREP(XIP_CTRL_INST_L_MASK,
+					 XIP_CTRL_INST_L0);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_INST_L_MASK,
+					  XIP_WRITE_CTRL_INST_L0);
+		break;
+	case 1:
+		ctrl->read |= XIP_CTRL_INST_EN_BIT
+			   |  FIELD_PREP(XIP_CTRL_INST_L_MASK,
+					 XIP_CTRL_INST_L8);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_INST_L_MASK,
+					  XIP_WRITE_CTRL_INST_L8);
+		break;
+	case 2:
+		ctrl->read |= XIP_CTRL_INST_EN_BIT
+			   |  FIELD_PREP(XIP_CTRL_INST_L_MASK,
+					 XIP_CTRL_INST_L16);
+		ctrl->write |= FIELD_PREP(XIP_WRITE_CTRL_INST_L_MASK,
+					  XIP_WRITE_CTRL_INST_L16);
+		break;
+	default:
+		LOG_ERR("Command length %d not supported", cmd_length);
+		return false;
+	}
```

Comment:
wondering if something like this could be done:

```
#define XIP_CTRL_INST_L_BYTES(n) n
```

since all these L macros expand to the actual `cmd_length` value. Then simplify a bit the code by removing some branches.